### PR TITLE
PFDR-158 - Reorganize chipmunk lib

### PR DESCRIPTION
**Pull request by botimer**
There are incremental, but meaningfully chunked commits here to repackage the stuff in `lib/`. The repository is not yet moved because there are specific challenges around its binding to both the Package model and the Bag concept. The Bag class is also not yet used at all because I want to discuss which pieces are bag-specific and which should comprise the general `Chipmunk::Package` interface.

I believe this is ready and safe to merge to master since it is all refactoring so far.

### CODE REVIEW

**Comment by aelkiss** [comment]
All looks fine to me. We probably need to tweak the rspec rubocop settings, though.

**Comment by botimer** [comment]
Rubocop is complaining about `Style/ClassAndModuleChildren` (configured to use nesting rather than compact declaration), but I have made conscious decisions about the three places where I used compact form.

I find it objectionable to keep indenting the main body of a file beyond the containment and declaration of the class/module of interest. This comes up with classes like `Chipmunk::Bag::Tag`, which is currently declared like so:

```
class Chipmunk::Bag
  class Tag
```

I originally had:

```
module Chipmunk
  class Bag::Tag
```

The latter has some merits, but I prefer the former. It is a detector for incidental dependencies in that it declares that this class operates within the scope of `Chipmunk::Bag`. If you needed to reach out to other parts of `Chipmunk`, the prefix with `Chipmunk::` signals the interdependency better. It also provides that siblings can be referenced without ceremony.

This is all opposed to the Rubocop default, which is to nest each level explicitly. I find that arbitrary, as far as formatting (files all having a different left margin and relocation causing gross indentation changes), and problematic in terms of introducing multiple lexical scopings in the same file, obscuring some design smells.

These are conscious design decisions and not "smelly" like line length, which I believe warrants case-by-case comment/disable in the file. I propose that we disable the `Style/ClassAndModuleChildren` cop, for these files at the least, in `.rubocop.yml`.

**Comment by aelkiss** [comment]
I concur. That cop has caused problems for us before.

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -3,6 +3,8 @@
 require "coveralls"
 Coveralls.wear!("rails")
 
+require "chipmunk"
```
Comment (malakai97) [inline spec/spec_helper.rb]: Requiring this here breaks test isolation, and obscures any issues with missing requires in the non-test code.

File: lib/chipmunk.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Chipmunk
+end
+
+require "semantic_logger"
```
[the remark this answers is not in the document]
Comment (botimer) [reply lib/chipmunk.rb]: I fully intend to use it "everywhere" we want to log. I'm okay with it being a "library dependency" (as would show up in gemspec/Gemfile use). Are you suggesting that we would require this in "many" files throughout `lib/`?

**Comment by botimer** [comment]
> The rubocop default is nested because it ensures the innermost child has full access to all the parents' scopes. In your example, `Chipmunk::Bag::Tag` will not be able to see constants defined in `Chipmunk`. The ability to change without touching indentation is nice, though.

I'm just not convinced that "all enclosing scopes" is a positive influence on design. Using bare names hurts context independence, but using them from from multiple different scopes amplifies it and masks cohesion issues -- too much for the benefit, IMO.

I've disabled the cop in this PR. Do you want to exclude these three specifically and take up a more structured decision for a general position?

File: spec/spec_helper.rb (hunk 1 above)
Comment (botimer) [reply spec/spec_helper.rb]: Hmmm... But requires are for the lifetime of the suite, and so, spec files are not genuinely isolated to begin with.

The theory here is that we should treat the "chipmunk library" like a gem. We `require "chipmunk"` and get the whole library. Granted, there are gems that have multiple modules or optional pieces, but I don't think there will be enough code (or downstream dependencies like drivers) here to warrant the performance/confusion tradeoff of partial loading now.

We `require "chipmunk"` in the services initializer in the app. That will be run in anything that would "boot" the app (including rake tasks with an 'environment' dependency). I think requiring it here is _more_ consistent and deterministic in the big picture than scattering loads across VM uptime.

To compare with Java, this "require it now" approach is more like adding a jar to the classpath. The imports for name resolution are entirely unrelated to loading. Doing incremental requires during runtime is more like `Class.forName`, which can be powerful for dynamic dependencies, but can also cause nondeterminism if not managed carefully.

**Comment by malakai97** [comment]
I'm not concerned about the scoping. I'd like to eventually see the broad `require`s removed. If that can be now, great; otherwise it doesn't block this change.

**Pull request merged**